helper/resource/testing_config.go

```diff
@@ -101,10 +101,12 @@ func testStep(
 	}
 
 	// And another after a Refresh.
-	state, err = ctx.Refresh()
-	if err != nil {
-		return state, fmt.Errorf(
-			"Error on follow-up refresh: %s", err)
+	if !step.Destroy || (step.Destroy && !step.PreventPostDestroyRefresh) {
```

Comment:
It probably is a JFDI approach to this problem.

I think the "right approach" ™️  would be (optionally) performing selective post-destroy refresh (i.e. exclude all or some data sources), but that would probably require more digging and more +/- LOC and I didn't want to end up in that 🐰  hole.

> I expect that in that case Terraform sould understand that the data source depends on the bucket and the object and would thus not try to refresh them when the resources are not present. > Unfortunately I expect that this will not work as desired ...

I can imagine the edge cases that aren't quite well covered yet which is exactly why I tried to keep data sources as isolated from resources as possible - after all that's going to be the common use case.

There is no reason why should user be attaching S3 object data source to an S3 object resource - they would use resource attributes instead. We should be testing common use cases. 😉 

i.e. I agree it should behave like you described, but that behaviour is not something we want to test in this specific context (s3-object-resource / s3-object-data-source).